### DISCUSSION TITLE
How to determine current WASM port?

### DISCUSSION THREAD

**Discussion by xperiandri** [unoplatform/uno#6382] [Q&A] [answered]
In order to pass a port to MSAL redirect callback, I need to determine the current port of running WASM app.
How to do that?

**Comment by carldebilly** [comment, marked as answer]
Hi @xperiandri, you may want to fix it using the `launchSettings.json` file.

Here's an example file doing that in the Uno project: https://github.com/unoplatform/uno/blob/master/src/SamplesApp/SamplesApp.Wasm/Properties/launchSettings.json

**Reply by xperiandri** [reply]
Isn't it possible to read that from the URL somehow? It depends on HTTP/HTTPS and may vary depending on deployment environment

**Reply by xperiandri** [reply]
I need to set redirect URI for MSAL according to the current URL

**Reply by carldebilly** [reply]
There's a feature for that in the `WebAuthenticationBroker`: https://github.com/unoplatform/uno/blob/c973d443faccce2778b5e81963b02a8adb46a8ea/src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.wasm.cs#L12_L22

Alternative, you can call this: https://docs.microsoft.com/en-us/uwp/api/windows.security.authentication.web.webauthenticationbroker.getcurrentapplicationcallbackuri?view=winrt-20348

**Reply by xperiandri** [reply]
`error Uno0001: Windows.Security.Authentication.Web.WebAuthenticationBroker.GetCurrentApplicationCallbackUri() is not implemented in Uno`

**Reply by xperiandri** [reply]
``` C#
#if __WASM__
Windows.Security.Authentication.Web.WebAuthenticationBroker.GetCurrentApplicationCallbackUri().AbsoluteUri
#else
```

**Reply by xperiandri** [reply]
I've even wrapped code into #if and it does not build

**Reply by xperiandri** [reply]
``` C#
#pragma warning disable Uno0001 // Uno type or member is not implemented
WebAuthenticationBroker
    .GetCurrentApplicationCallbackUri()
    .AbsoluteUri
    .Replace("authentication-callback", "authentication/login-callback.htm")
#pragma warning restore Uno0001 // Uno type or member is not implemented
```
Only this helped

**Reply by xperiandri** [reply]
@carldebilly, @jeromelaban, probably you forgot to remove `Uno0001` for supported platforms

**Reply by xperiandri** [reply]
@carldebilly that is strange that for MSAL and for AuthenticationBroker itself the callback URI is different

**Reply by xperiandri** [reply]
https://github.com/unoplatform/uno/blob/c973d443faccce2778b5e81963b02a8adb46a8ea/src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.wasm.cs#L17

Something like?

**Reply by carldebilly** [reply]
@xperiandri, it's very stange you're getting this error (`Uno0001`). I checked on my side and it works.

It has been implemented in Uno.UI version 3.6+ so you need at least this version to have this feature.

**Reply by xperiandri** [reply]
The package I use is 3.8.13
And I have that warning
![image](https://user-images.githubusercontent.com/2365592/125981790-d876dd26-e854-4acf-ba0d-c758d3a8e79a.png)

**Reply by carldebilly** [reply]
I did the same of you, on a new project with the same Uno version and I'm not getting this warning.  Right now I have no explanation for this problem.  @jeromelaban, any idea?

Also, I notice you're replacing the output for the right path. Use this trick instead:

``` csharp
// could be set globally, no need to set this each time
WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultCallbackPath = "/authentication/login-callback.htm"; 

// should produce the right return url...
var redirectUri = Windows.Security.Authentication.Web.WebAuthenticationBroker.GetCurrentApplicationCallbackUri();
```

**Reply by xperiandri** [reply]
Thanks!

**Reply by xperiandri** [reply]
Try with this
``` XML
  <PropertyGroup Condition="'$(Configuration)'=='Release'">
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <WarningsAsErrors />
  </PropertyGroup>
```
on WASM

**Reply by carldebilly** [reply]
Same... it's not triggerring the Uno0001 analyzer here.

**Reply by xperiandri** [reply]
Interesting, why can it happen on my project configuration

**Reply by jeromelaban** [reply]
This API is not marked as not implemented, it could be that your project is referencing an older version of uno. When you "go to definition" on `GetCurrentApplicationCallbackUri`, what's the path to the dll ?

**Reply by xperiandri** [reply]
``` C#
#region Assembly Uno, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null
// C:\Users\Andrii\.nuget\packages\uno.ui\3.8.13\lib\netstandard2.0\Uno.dll
#endregion
```

**Reply by carldebilly** [reply]
1. Close all instances of Visual Studio
2. Flush all you Nuget packages:
   ```
   C:\sources\myproject> nuget locals all -clear
   ```
3. Clean your local repo: (assuming you are using git and all your relevant changes are staged)
   ```
   C:\sources\myproject> git clean -fdx
   ```

Build again.

**Reply by jeromelaban** [reply]
Ah I see. It's the `WebAuthentiationBroker` class that's marked as not implemented, not the method.

https://github.com/unoplatform/uno/blob/56d3ba98ee6cab055dc8d7c4941ce4494e20e946/src/Uno.UWP/Generated/3.0.0.0/Windows.Security.Authentication.Web/WebAuthenticationBroker.cs#L6

You can add the `#pragma warning disable` for now

**Reply by xperiandri** [reply]
Thanks, I did that.
I will try the solution above tomorrow on broadband internet

**Reply by jeromelaban** [reply]
It was actually already fixed, should be released in 3.9: https://github.com/unoplatform/uno/issues/6252
